File: app/src/main/java/com/stackanswer/source/datasource/LocalShowDataSource.java
package com.stackanswer.source.datasource;

import androidx.paging.DataSource;

import com.stackanswer.source.local.room.showfavorite.ShowFavorite;
import com.stackanswer.source.local.room.showfavorite.ShowFavoriteDao;

public class LocalShowDataSource {

    private static LocalShowDataSource INSTANCE;
    private final ShowFavoriteDao mShowFavoriteDao;

    private LocalShowDataSource(ShowFavoriteDao mShowFavoriteDao) {
        this.mShowFavoriteDao = mShowFavoriteDao;
    }

    public static LocalShowDataSource getInstance(ShowFavoriteDao academyDao) {
        if (INSTANCE == null) {
            INSTANCE = new LocalShowDataSource(academyDao);
        }
        return INSTANCE;
    }

    public DataSource.Factory<Integer, ShowFavorite> getAllShow() {
        return mShowFavoriteDao.getAllShow();
    }

    public void update(ShowFavorite courses) {
        mShowFavoriteDao.update(courses);
    }

    public void delete(ShowFavorite courses) {
        mShowFavoriteDao.delete(courses);
    }

    public void insert(ShowFavorite courses) {
        mShowFavoriteDao.insert(courses);
    }

    public void deleteAllShow() {
        mShowFavoriteDao.deleteAllShow();
    }
}

File: app/src/main/java/com/stackanswer/source/datasource/kotlin/LocalShowDataSource.kt
package com.stackanswer.source.datasource.kotlin

import com.stackanswer.source.local.room.show.ShowDao
import com.stackanswer.source.local.room.show.ShowPopular
import io.reactivex.Flowable

class LocalShowDataSource private constructor(private val tourismDao: ShowDao) {

    companion object {
        private var instance: LocalShowDataSource? = null

        fun getInstance(tourismDao: ShowDao): LocalShowDataSource =
            instance ?: synchronized(this) {
                instance ?: LocalShowDataSource(tourismDao)
            }
    }

    fun getAllTourism(): Flowable<List<ShowPopular>> = tourismDao.getAllTourism()

    fun insertTourism(tourismList: List<ShowPopular>) =
        tourismDao.insertTourism(tourismList)

//    fun getFavoriteTourism(): Flowable<List<ShowPopular>> = tourismDao.getFavoriteTourism()

//    fun setFavoriteTourism(tourism: ShowPopular, newState: Boolean) {
//        tourism.isFavorite = newState
//        tourismDao.updateFavoriteTourism(tourism)
//    }
}